### ISSUE TITLE
Parse Error: Expected HTTP/

### ISSUE THREAD

**Issue by Ankcorn** [httptoolkit/mockttp#38]
On mockttp 0.20.0, 0.20.1, and 0.20.2 I get a weird http parse error. I think it is something to do with response encoding.

Node version is 12.18.0

tested the request with request-promise, axios, and fetch with the same results each time

```
{
  "config": {
    "url": "https://localhost:1234/oauth/access_token.php",
    "method": "post",
    "data": {
      "_overheadLength": 432,
      "_valueLength": 150,
      "_valuesToMeasure": [],
      "writable": false,
      "readable": true,
      "dataSize": 0,
      "maxDataSize": 2097152,
      "pauseStreams": true,
      "_released": true,
      "_streams": [],
      "_currentStream": null,
      "_insideLoop": false,
      "_pendingNext": false,
      "_boundary": "--------------------------229731192039541199026626",
      "_events": {},
      "_eventsCount": 1
    },
    "headers": {
      "Accept": "application/json, text/plain, */*",
      "Content-Type": "multipart/form-data; boundary=--------------------------229731192039541199026626",
      "User-Agent": "axios/0.19.2"
    },
    "transformRequest": [
      null
    ],
    "transformResponse": [
      null
    ],
    "timeout": 0,
    "xsrfCookieName": "XSRF-TOKEN",
    "xsrfHeaderName": "X-XSRF-TOKEN",
    "maxContentLength": -1
  },
  "code": "HPE_INVALID_CONSTANT"
}
```

**Comment by pimterry** [comment]
This error comes directly from Node's HTTP parser, try searching for HPE_INVALID_CONSTANT in here: https://github.com/nodejs/http-parser/blob/master/http_parser.c. 

It is possible that Mockttp somehow interferes with the data that node's receiving, but it's unlikely. This is probably a real problem with the request itself (or the response, if you're proxying this traffic to a real server).

Can you share the code that's sending the request?

You could also try making the same request manually through [HTTP Toolkit](https://httptoolkit.tech/), which will show the same error but hopefully with more context, or [Wireshark](https://www.wireshark.org/) (if it's not HTTPS).

**Comment by Ankcorn** [comment]
I see, the code still works when you are making requests to the actual http server so I think it could be that unlikely scenario 😨 

On mockttp 0.19.3 the code works.

The client we are calling the api with looks like 
```javascript
const axios = require('axios');

function xApi(key) {
	const apiKey = key;

	async function create(data) {
		return (await axios.post(`${process.env.X_SERVICE_URL}`, data, {
			headers: { 'x-api-key': apiKey }
		})).data;
	}

	async function update(data, id) {
		return (await axios.put(`${process.env.X_SERVICE_URL}/internal/${id}`, data, {
			headers: { 'x-api-key': apiKey }
		})).data;
	}

	return {
		create,
		update
	};
}

module.exports = xApi;
```

I am downloading the http toolkit now to get more info

**Comment by pimterry** [comment]
Ah, interesting!

What does your Mockttp mocking code for this request look like?

One guess: are you mocking the response headers here? What happens if you add `"Transfer-Encoding": "chunked"` to those?

In Mockttp < 0.20, some extra default headers (like transfer-encoding, and various others) were silently added automatically, in addition to headers you manually specified. This wasn't great - mocks weren't really accurate to their configuration, and it wasn't possible to test some invalid response cases.

In Mockttp 0.20+, if you specify any `headers` property for your mock then you take responsibility for specifying _all_ response headers. To make your response body valid, you might need either a chunked encoding or a content-length valid. That sounds like it might cause this kind of thing.

**Comment by Ankcorn** [comment]
We do not specify headers but we are using .thenJSON to send our responses.
```javascript
       const { getLocal } = require('mockttp');
        const certs = {
	        https: {
		        keyPath: './testCA.key',
		        certPath: './testCA.pem'
	        }
        };
        const xServer = getLocal(certs);
	const _xRequests = await xServer.post('/')
			.thenJSON(200, { id: '🐶', xNumber: 888, secretToken: 'shhhh' });
```

**Comment by Ankcorn** [comment]
https://github.com/nodejs/http-parser/issues/517

Here is the issue in http parser that could be causing this

**Comment by Ankcorn** [comment]
With that header I get this error

```javascript
{
    "errorMessage": "Parse Error: Content-Length can't be present with Transfer-Encoding",
    "errorType": "Error",
    "stackTrace": [
        "Error: Parse Error: Content-Length can't be present with Transfer-Encoding",
        "    at Socket.socketOnData (_http_client.js:476:22)",
        "    at Socket.emit (events.js:310:20)",
        "    at Socket.EventEmitter.emit (domain.js:482:12)",
        "    at addChunk (_stream_readable.js:286:12)",
        "    at readableAddChunk (_stream_readable.js:268:9)",
        "    at Socket.Readable.push (_stream_readable.js:209:10)",
        "    at TCP.onStreamRead (internal/stream_base_commons.js:186:23)"
    ]
}

I also noticed its only happening with axios not node-fetch. Unless you have a better suggestion I can raise the issue there.


TBH I don't fully understand whats going on 🤷‍♂️

**Comment by pimterry** [comment]
Ok, I think I've worked out the root cause of this, it is indeed a Mockttp bug. This code by itself fails with the "Parse Error: Expected HTTP/" error you describe:

```js
const mockttp = require('mockttp');
const axios = require('axios').default;

const server = mockttp.getLocal();

server.start().then(async() => {
    await server.get('/')
        .thenJSON(200, { id: '🐶', xNumber: 888, secretToken: 'shhhh' });

    await axios.get(server.url);
});
```

If you replace the emoji in `id` with `"123"` it doesn't fail any more.

I think what's happening here is that `thenJSON` is calculating the content length for you ([here](https://github.com/httptoolkit/mockttp/blob/master/src/rules/mock-rule-builder.ts#L303)), assuming everything is ASCII, and that's causing errors with unicode content like this.

Then because the content length is wrong, some of the content is treated by the HTTP client as part of a subsequent unexpected response, except it's really the end of the response body, so it doesn't parse.

Make sense? In the short term you can fix it by avoiding unicode, or specifying your own `headers` property for `thenJSON` with a corrected `Content-Length` header value. I'll look into it and get a proper fix for you though, shouldn't take too long.

**Comment by pimterry** [comment]
Ok, I think the version I've just pushed now [fixes](https://github.com/httptoolkit/mockttp/commit/9f6540a53c4b10f30cab934b0af06a1dd628e959) this. If you've got a minute, could you please:

* Clone this repo, and in that folder:
	* `npm install`, `npm run build`, `npm link`
* In your project:
	* `npm link mockttp`
	* Run your tests, and see if this fixes your issue

**Comment by Ankcorn** [comment]
Nooo emoji life why did you cause me pain 😂😂😂😂😂


Thanks so much for this, I will test it out first thing in the morning 🥰

**Comment by Ankcorn** [comment]
It works! thank you :)

**Comment by pimterry** [comment]
Great! Thanks for testing that. I'll put together a proper release in the next day or so, so you can use this for real, watch this space.

**Comment by pimterry** [comment]
This was just published as v0.20.3 :tada:.

Thanks for reporting this. Hope that sorts things out for you, let me know if you have any other trouble.
